internal/locate/region_cache.go

```diff
@@ -1240,25 +1240,30 @@ func (c *RegionCache) BatchLocateKeyRanges(bo *retry.Backoffer, keyRanges []kv.K
 				keyRange.StartKey = lastRegion.EndKey()
 			}
 		}
-		for {
-			// TODO: find all the cached regions in the range.
-			// now we only check if the region is cached from the lower bound, if there is a uncached hole in the middle,
-			// we will load the rest regions even they are cached.
-			r := c.tryFindRegionByKey(keyRange.StartKey, false)
-			lastRegion = r
-			if r == nil {
-				// region cache miss, add the cut range to uncachedRanges, load from PD later.
-				uncachedRanges = append(uncachedRanges, pd.KeyRange{StartKey: keyRange.StartKey, EndKey: keyRange.EndKey})
+		// TODO: find all the cached regions in the range.
+		// now we only check if the region is cached from the lower bound, if there is a uncached hole in the middle,
+		// we will load the rest regions even they are cached.
+		batchRegionInCache, err := c.scanRegionsFromCache(bo, keyRange.StartKey, keyRange.EndKey)
+		if err != nil {
+			return nil, err
+		}
+		hit := false
+		for _, r := range batchRegionInCache {
+			if !r.Contains(keyRange.StartKey) { // uncached hole, load the rest regions
```

Comment:
I don't understand this code. If `[keyRange.StartKey, keyRange.Endkey)` contains multiple regions, in my understanding only the first region can contain `keyRange.StartKey`?

Reply:
After processing a region, I set keyRange.StartKey = r.endkey(). So the next region also can contain the  keyRange.StartKey.